metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -78,42 +301,221 @@ inline bool fetch_and_unset_bit(uint64_t *v,
     }
 }
 
-inline bool fetch_bit(const uint64_t *v,
-                      uint64_t i,
-                      bool atomic = false,
-                      int mo = __ATOMIC_SEQ_CST) {
+inline bool fetch_bit(const uint64_t *v, uint64_t i, bool atomic, int mo) {
```

Comment:
It was done this way because we originally thought that the atomic operations would introduce too much overhead in the traversal operations, so we wanted to make them variable so that a non-atomic version would be used in the single threaded case. However, it turned out in the end that the overhead is negligible relative to the rest of the operations and we figured that the branch predictor would learn that the same branch is always taken here.

Reply:
I am not sure if branch prediction is also able to amortize the effects of the jmp. Maybe it is. But it still seems to me that we can easily avoid that overhead, so why don't we? :)

Reply:
I'll put a TODO so we can handle this in a different pull request later, since it'll require changing more stuff and merging might get messy.

Reply:
Doesn't jmp cost zero if branch prediction works?

Do you know if there is any overhead of using atomic operations at all? Suppose you have a single thread, so no atomicity is needed, but you still call these atomic operations.

Reply:
I don't know what the cost is if the prediction works, the documentation was unclear. It seemed to indicate that the overhead of jmp is still there, but maybe because it's done ahead of time is negligible dunno.

Regarding atomic operations: as far as I know, atomic operations simply block the cache line to ensure atomicity, so they cause an overhead if there is another processor wanting to use that cache line. No idea what that involves in practice.

Reply:
When I was parallelizing the sequence extraction method, I benchmarked it with and without atomic operations and the overhead was negligible (<1%), so the other graph traversal operations dominate.

Reply:
But I've read that the overhead can be 10x in cases of high contention.

Reply:
Yeah, I agree. In graph traversal we cal always use atomic calls. But I am not sure we should remove this feature from the interface completely. Maybe it will be useful in other cases? But I don’t know.

Reply:
Regarding branch prediction: there is some overhead because the if statement still has to be evaluated eventually for every call, but there won't be any branch misprediction overhead.